ZEngine/ZEngine/Core/Maths/Matrix.h

```diff
@@ -0,0 +1,483 @@
+        }
+    };
+
+    template <typename T>
+    struct Mat2 : public Matrix<T, 2, 2>
+    {
+
+        Mat2()
+        {
+            memset(this->m_data, 0, sizeof(this->m_data));
+        }
+
+        Mat2(T m00, T m01, T m10, T m11)
+        {
+            this->m_data[0] = m00;
+            this->m_data[1] = m01;
+            this->m_data[2] = m10;
+            this->m_data[3] = m11;
+        }
+
+        Mat2(const Matrix<T, 2, 2>& other)
+        {
+            for (size_t i = 0; i < 4; ++i)
+                this->m_data[i] = other.m_data[i];
+        }
+
+        T determinant() const
+        {
+            return this->m_data[0] * this->m_data[3] - this->m_data[1] * this->m_data[2];
+        }
+
+        Mat2& operator+=(const Mat2& other)
+        {
+            for (size_t i = 0; i < 4; i++)
+            {
+                this->m_data[i] += other.m_data[i];
+            }
+            return *this;
+        }
+        Mat2& operator-=(const Mat2& other)
+        {
+            for (size_t i = 0; i < 4; i++)
+            {
+                this->m_data[i] -= other.m_data[i];
+            }
+            return *this;
+        }
+        Mat2 operator+(const Mat2& other)
+        {
+            Mat2 result  = *this;
+            result      += other;
+            return result;
+        }
+        Mat2 operator-(const Mat2& other)
+        {
+            Mat2 result  = *this;
+            result      -= other;
+            return result;
+        }
+        Mat2 operator*(T scalar)
+        {
+            Mat2 result = *this;
+            for (size_t i = 0; i < 4; i++)
+            {
+                result.m_data[i] *= scalar;
+            }
+            return result;
+        }
+        Mat2& operator*=(T scalar)
+        {
+            for (size_t i = 0; i < 4; ++i)
+            {
+                this->m_data[i] *= scalar;
+            }
+            return *this;
+        }
+
+        Mat2 operator/(T scalar) const
+        {
+            Mat2 result = *this;
+            scalar      = 1.0f / scalar;
+            for (size_t i = 0; i < 4; i++)
+            {
+                result.m_data[i] *= scalar;
+            }
+            return result;
+        }
+
+        Mat2& operator/=(T scalar)
+        {
+            scalar = 1.0f / scalar;
+            for (size_t i = 0; i < 4; ++i)
+            {
+                this->m_data[i] *= scalar;
+            }
+            return *this;
+        }
+        Mat2 Inverse()
+        {
+            T det = this->determinant();
+            ZENGINE_VALIDATE_ASSERT(det != 0, "Matrix is singular and cannot be inverted");
+
+            T invDet = 1 / det;
+
+            return Mat2<T>(this->m_data[3] * invDet, -this->m_data[1] * invDet, -this->m_data[2] * invDet, this->m_data[0] * invDet);
+        }
+    };
+
+    template <typename T>
+    struct Mat3 : public Matrix<T, 3, 3>
+    {
+
+        Mat3()
+        {
+            memset(this->m_data, 0, sizeof(this->m_data));
```

Comment:
+ we have secure_memset(....)
+ it should be sizeof(T) * R * C, not sizeof(m_data)